Record most recent snapshot policy success/failure

gwbrown opened the pull request:
Keeping a record of the results of the successes and failures will aid
troubleshooting of policies and make users more confident that their
snapshots are being taken as expected.

This is the first step toward writing history in a more permanent
fashion.

This is still a little rough (it doesn't yet convert timestamps to human-readable strings in the API response), but it's a first cut.

Relates to https://github.com/elastic/elasticsearch/issues/38461

elasticmachine commented:
Pinging @elastic/es-core-features

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/action/GetSnapshotLifecycleAction.java:
```diff
@@ -146,22 +152,34 @@ public boolean equals(Object obj) {
+            this.modifiedDate = policyMetadata.getModifiedDate();
+            this.lastSuccessDate = policyMetadata.getLastSuccessDate();
+            this.lastFailureDate = policyMetadata.getLastFailureDate();
+            this.lastFailureInfo = policyMetadata.getLastFailureInfo();
```
I'm also thinking, we *may* want to separate this into two different objects so they are clearly defined and separated, something like:


```
{
    ... policy json etc ...
    "last_success": {
        "snapshot_name": "production-snapshot-2019-03-15-kwno8wplqfaswa5rsbvdvg",
        "time": 1553813657,
        "time_string": "2019-03-15.21:09:00-blah"
    },
    "last_failure": {
        "snapshot_name": "production-snapshot-2018-12-02-aoeu890gaoeuhaoeu098as",
        "time": 1553629109,
        "time_string": "2018-12-02.19:23:01",
        "failure": "java.lang.NotCoolEnoughException: you are not cool enough to take snapshots\n\torg.try.to.take.the.snapshot.DoIt:129\n\t...etc..."
    }
}
```

I'm trying to think of the easiest thing to consume, I think having only dates and an error (which only pertains to the failure, rather than the success, will be a little confusing.

What do you think about output like this?

gwbrown replied:
I had considered doing that originally, and only didn't since there were only a couple fields. If we're going to include more information (which I agree would be good, I didn't think of the snapshot name) then yes, I'll switch these to object fields instead.

gwbrown commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/SnapshotLifecyclePolicyMetadata.java:
```diff
@@ -150,4 +229,57 @@ public String toString() {
+        }
+
+        public SnapshotLifecyclePolicyMetadata build() {
+            Objects.requireNonNull(modifiedDate, "modifiedDate must be set");
```
`policy` is null-checked in the constructor - I can move that check here, I only split them because I originally did it in the constructor but a null `modifiedDate` would throw an NPE at the time it gets passed into the constructor. I'll do all the checks in one place.

Good point on `headers`, I'll do that too.

gwbrown commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTask.java:
```diff
@@ -78,4 +94,80 @@ public void onFailure(Exception e) {
+        }
+
+        private String exceptionToString() throws IOException {
+            if (exception.isPresent()) {
```
Only partially true - I'm changing this to use try-with-resources as in your example, but it can't be done with `Optional.map()` because that doesn't play nice with things that can throw exceptions.

Also, I copied this a bit from ILM and I think we should do this here, too:
https://github.com/elastic/elasticsearch/blob/647e2256984abd44d5d0820bc742a0af2061a2b8/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleRunner.java#L334-L343

gwbrown commented:
Thanks for the thorough review @dakrone! I think I've addressed all the feedback except for the tests for `SnapshotLifecyclePolicyItem`, which I'll work on as well, but can you take a look at the new changes?

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTask.java:
```diff
@@ -78,4 +95,82 @@ public void onFailure(Exception e) {
+
+            assert snapMeta != null : "this should never be called while the snapshot lifecycle cluster metadata is null";
+            if (snapMeta == null) {
+                logger.error("failed to record snapshot [{}] for snapshot policy [{}]: snapshot lifecycle metadata is null",
```
We might want to log the snapshot name also so the user can go check it out manually

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTask.java:
```diff
@@ -78,4 +95,82 @@ public void onFailure(Exception e) {
+            Map<String, SnapshotLifecyclePolicyMetadata> snapLifecycles = new HashMap<>(snapMeta.getSnapshotConfigurations());
+            SnapshotLifecyclePolicyMetadata policyMetadata = snapLifecycles.get(policyName);
+            if (policyMetadata == null) {
+                logger.warn("failed to record snapshot [{}] for snapshot policy [{}]: policy not found",
```
We might want to log the snapshot name also so the user can go check it out manually

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTask.java:
```diff
@@ -78,4 +95,82 @@ public void onFailure(Exception e) {
+
+        @Override
+        public void onFailure(String source, Exception e) {
+            logger.error("failed to record snapshot policy execution status [{}]: {}", source, e);
```
Same here about maybe logging the name of the snapshot

gwbrown commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/SnapshotLifecyclePolicyMetadata.java:
```diff
@@ -140,14 +207,71 @@ public boolean equals(Object obj) {
         // should not emit them in case it accidentally gets logged.
-        return super.toString();
+//        return super.toString();
+        return Strings.toString(this); //NOCOMMIT
```
I did this while debugging and put NOCOMMIT here to make sure I remembered to remove it! Unfortunately it looks like that doesn't work at the moment. Thanks for catching it!

gwbrown commented:
I'll do those in a follow-up PR, although I'm going to update this one to fix the integration test so it won't break with #40654.

gwbrown commented:
@dakrone I've updated the test to use an index that doesn't exist and `ignore_unavailable` to trigger a failure, as you suggested.
